Lib/bdb.py

```diff
@@ -309,10 +359,26 @@ def set_step(self):
                 caller_frame.f_trace = self.trace_dispatch
         self._set_stopinfo(None, None)
 
+    def set_stepinst(self, frame):
+        """Stop after one opcode."""
+        # Issue #13183: pdb skips frames after hitting a breakpoint and running
+        # step commands.
+        # Restore the trace function in the caller (that may not have been set
+        # for performance reasons) when returning from the current frame.
+        if self.frame_returning:
+            caller_frame = self.frame_returning.f_back
+            if caller_frame and not caller_frame.f_trace:
+                caller_frame.f_trace = self.trace_dispatch
```

Comment:
Consider extracting this piece into a function (e.g. `self.restore_trace_function()`) since it is used twice: in `set_step` and here